pmd-java/src/test/java/net/sourceforge/pmd/typeresolution/TypeInferenceTest.java

```diff
@@ -0,0 +1,206 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.typeresolution;
+
+
+import static net.sourceforge.pmd.lang.java.typeresolution.typeinterference.InferenceRuleType.CONTAINS;
+import static net.sourceforge.pmd.lang.java.typeresolution.typeinterference.InferenceRuleType.EQUALITY;
+import static net.sourceforge.pmd.lang.java.typeresolution.typeinterference.InferenceRuleType.LOOSE_INVOCATION;
+import static net.sourceforge.pmd.lang.java.typeresolution.typeinterference.InferenceRuleType.SUBTYPE;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+import java.util.List;
+import java.util.Map;
+
+import org.junit.Test;
+
+
+import net.sourceforge.pmd.lang.java.typeresolution.typedefinition.JavaTypeDefinition;
+import net.sourceforge.pmd.lang.java.typeresolution.typeinterference.Bound;
+import net.sourceforge.pmd.lang.java.typeresolution.typeinterference.BoundOrConstraint;
+import net.sourceforge.pmd.lang.java.typeresolution.typeinterference.Constraint;
+import net.sourceforge.pmd.lang.java.typeresolution.typeinterference.InferenceRuleType;
+import net.sourceforge.pmd.lang.java.typeresolution.typeinterference.Variable;
+
+public class TypeInferenceTest {
+    private JavaTypeDefinition number = JavaTypeDefinition.forClass(Number.class);
+    private JavaTypeDefinition integer = JavaTypeDefinition.forClass(Integer.class);
+    private JavaTypeDefinition primitiveInt = JavaTypeDefinition.forClass(int.class);
+    private JavaTypeDefinition generic = JavaTypeDefinition.forClass(Map.class, number, integer);
+    private Variable a = new Variable();
+    private Variable b = new Variable();
+
+    @Test
+    public void testEqualityReduce() {
+
+        // If S and T are proper types, the constraint reduces to true if S is the same as T (§4.3.4), and false
+        // otherwise.
+        assertTrue(new Constraint(number, number, EQUALITY).reduce().isEmpty());
+        assertEquals(new Constraint(number, integer, EQUALITY).reduce(), null);
+
+        // Otherwise, if S or T is the null type, the constraint reduces to false. TODO
```

Comment:
add the test, if not currently implemented, just add an `@Ignore` to it, but we should avoid TODOs when we are able to